File: contrib/MITRE/examples/double/java/Multiply.java
/*
  This file (c) Copyright 1998 - 2002 The MITRE Corporation
  
  This file is part of the Galaxy Communicator system. It is licensed
  under the conditions described in the file LICENSE in the root 
  directory of the Galaxy Communicator system.
*/

/**
 * $Id: Multiply.java,v 1.11 2002/03/21 16:15:29 wohlever Exp $
 */
import galaxy.lang.*;
import galaxy.server.*;
import galaxy.server.ui.ServerUI;
import java.net.*;
import java.io.*;

import galaxy.util.Logger;

public class Multiply extends galaxy.server.Server {

    int factor = 1;
 
    public Multiply(MainServer ms, Socket s) 
	throws IOException {
	super(ms,s);
	initSignatures();
    } 
    
    private void initSignatures() {
    	SigEntry[] foo = {new SigEntry(":int",GalaxyObject.GAL_INT, Signature.GAL_KEY_ALWAYS)};
	addSignature(new Signature("multiply", foo , Signature.GAL_OTHER_KEYS_NEVER, foo, Signature.GAL_OTHER_KEYS_NEVER, Signature.GAL_REPLY_PROVIDED));
    }
    
    public GFrame serverOpMultiply(GFrame fr) {
	int i = fr.getInt(":int").intValue();
	// Do some error checking...
	if ( (Integer.MAX_VALUE/i) < factor) {
	    
	    try {
		getCurrentEnvironment().error("multiply would overflow MAXINT");
	    } catch (Exception ex) {
		logErrorMessage("Caught exception while sending error message: " + ex.toString(), ex);
	    }
	    return (GFrame)null;
	}
	fr.setProperty(":int", i * factor);
	return fr;
    }

    public void serverOpReinitialize(GFrame fr) {
	int f = fr.getInt(":factor").intValue();
	if (f != 0)
	    this.factor = f;
    }

    public static void main(String[] args) {
	
	/*
	ServerUI ui = new ServerUI(args) {
		protected void init() {
		    serverClassName = "Multiply";
		    port = 2900;
		    serverName = "multiply";
		}};
	*/
	
	MainServer mainServer = new MainServer("multiply", args, 2900);
	mainServer.setServerClassName("Multiply");
	try {
	    mainServer.start();
	} catch(Exception ex) {
	    Logger.getLogger().logErrorMessage("Caught exception while staring main server: " + ex.toString(), ex);
	}
    }
}

File: contrib/MITRE/examples/double/python/multiply.py
# This file (c) Copyright 1998 - 2002 The MITRE Corporation
# 
# This file is part of the Galaxy Communicator system. It is licensed
# under the conditions described in the file LICENSE in the root 
# directory of the Galaxy Communicator system.

import os, sys, string, math, random

sys.path.insert(0, os.path.join(os.environ["GC_HOME"],
				"contrib", "MITRE", "templates"))

import GC_py_init

import Galaxy, GalaxyIO

Factor = 1

import exceptions

MultiplyError = "MultiplyError"

def Multiply(env, dict):
    try:
        res = dict[":int"] * Factor
        if (type(dict[":int"]) is type(0)) and \
           (type(res) is type(0L)):
            # Overflow.
            raise MultiplyError, "multiply would overflow MAXINT"
        else:
            return {":int": res}
    except exceptions.OverflowError:
        raise MultiplyError, "multiply would overflow MAXINT"    

def Welcome(env, dict):
    global Factor
    try:
	Factor = dict[":factor"]
    except: pass
    return None

def main():
    s = GalaxyIO.Server(sys.argv, "multiply", default_port = 2900)
    s.AddDispatchFunction("multiply", Multiply,
			  [[[":int", Galaxy.GAL_INT, Galaxy.GAL_KEY_ALWAYS]],
			   Galaxy.GAL_OTHER_KEYS_NEVER, 
			   Galaxy.GAL_REPLY_PROVIDED,
			   [[":int", Galaxy.GAL_INT, Galaxy.GAL_KEY_ALWAYS]],
			   Galaxy.GAL_OTHER_KEYS_NEVER])
    s.AddDispatchFunction("reinitialize", Welcome)
    s.RunServer()

main()
